### PR TITLE
service/elasticache: use waiter and finder function pattern

### PR DESCRIPTION
**Pull request by gdavison**
Refactors `aws_elasticache_cluster`, `aws_elasticache_replication_group`, `data.aws_elasticache_cluster`, and `data.aws_elasticache_replication_group` to use waiter and finder functions.

Relates #14959

Release note for [CHANGELOG](https://github.com/hashicorp/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
NONE
```

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccAWSElasticacheCluster_\|TestAccAWSElasticacheReplicationGroup_\|TestAccAWSDataElasticacheCluster_\|TestAccDataSourceAwsElasticacheReplicationGroup_'

--- PASS: TestAccAWSElasticacheReplicationGroup_clusteringAndCacheNodesCausesError (66.55s)
--- PASS: TestAccAWSElasticacheCluster_Memcached_FinalSnapshot (71.95s)
--- PASS: TestAccAWSElasticacheCluster_NumCacheNodes_Redis (74.25s)
--- PASS: TestAccAWSDataElasticacheCluster_basic (737.15s)
--- PASS: TestAccAWSElasticacheReplicationGroup_enableSnapshotting (986.85s)
--- PASS: TestAccAWSElasticacheReplicationGroup_updateDescription (987.69s)
--- PASS: TestAccAWSElasticacheReplicationGroup_useCmkKmsKeyId (1090.98s)
--- PASS: TestAccAWSElasticacheCluster_NumCacheNodes_IncreaseWithPreferredAvailabilityZones (1229.45s)
--- PASS: TestAccAWSElasticacheCluster_NodeTypeResize_Memcached (1311.80s)
--- PASS: TestAccAWSElasticacheReplicationGroup_enableAuthTokenTransitEncryption (1396.08s)
--- PASS: TestAccAWSElasticacheReplicationGroup_redisClusterInVpc2 (1420.73s)
--- PASS: TestAccAWSElasticacheReplicationGroup_updateParameterGroup (1485.93s)
--- PASS: TestAccAWSElasticacheCluster_EngineVersion_Redis (1425.75s)
--- PASS: TestAccAWSElasticacheCluster_ReplicationGroupID_SingleReplica (1500.06s)
--- PASS: TestAccAWSElasticacheCluster_NodeTypeResize_Redis (1516.86s)
--- PASS: TestAccAWSElasticacheCluster_ReplicationGroupID_MultipleReplica (1522.45s)
--- PASS: TestAccAWSElasticacheCluster_ReplicationGroupID_AvailabilityZone (1589.59s)
--- PASS: TestAccAWSElasticacheCluster_EngineVersion_Memcached (1528.96s)
--- PASS: TestAccAWSElasticacheReplicationGroup_enableAtRestEncryption (1644.84s)
--- PASS: TestAccAWSElasticacheCluster_AZMode_Redis (662.08s)
--- PASS: TestAccAWSElasticacheCluster_AZMode_Memcached (648.89s)
--- PASS: TestAccAWSElasticacheReplicationGroup_updateMaintenanceWindow (1987.14s)
--- PASS: TestAccDataSourceAwsElasticacheReplicationGroup_NonExistent (10.49s)
--- PASS: TestAccAWSElasticacheCluster_multiAZInVpc (783.12s)
--- PASS: TestAccAWSElasticacheReplicationGroup_NumberCacheClusters (2109.38s)
--- PASS: TestAccAWSElasticacheCluster_vpc (647.27s)
--- PASS: TestAccAWSElasticacheCluster_Port_Redis_Default (521.35s)
--- PASS: TestAccAWSElasticacheReplicationGroup_Uppercase (648.51s)
--- PASS: TestAccAWSElasticacheCluster_ParameterGroupName_Default (673.04s)
--- PASS: TestAccAWSElasticacheReplicationGroup_vpc (789.57s)
--- PASS: TestAccAWSElasticacheCluster_Engine_Memcached (635.77s)
--- PASS: TestAccAWSElasticacheReplicationGroup_multiAzInVpc (917.38s)
--- PASS: TestAccAWSElasticacheReplicationGroup_basic (852.65s)
--- PASS: TestAccAWSElasticacheCluster_Engine_Redis (704.39s)
--- PASS: TestAccAWSElasticacheReplicationGroup_FinalSnapshot (1025.95s)
--- PASS: TestAccAWSElasticacheCluster_Redis_FinalSnapshot (854.12s)
--- PASS: TestAccAWSElasticacheReplicationGroup_NumberCacheClusters_Failover_AutoFailoverDisabled (1541.39s)
--- PASS: TestAccAWSElasticacheCluster_NumCacheNodes_Increase (1074.42s)
--- PASS: TestAccAWSElasticacheCluster_snapshotsWithUpdates (716.36s)
--- PASS: TestAccAWSElasticacheCluster_Port (649.82s)
--- PASS: TestAccAWSElasticacheReplicationGroup_NumberCacheClusters_Failover_AutoFailoverEnabled (2048.20s)
--- PASS: TestAccDataSourceAwsElasticacheReplicationGroup_ClusterMode (1130.27s)
--- PASS: TestAccDataSourceAwsElasticacheReplicationGroup_basic (907.52s)
--- PASS: TestAccAWSElasticacheReplicationGroup_ClusterMode_NonClusteredParameterGroup (827.12s)
--- PASS: TestAccAWSElasticacheReplicationGroup_ClusterMode_Basic (861.43s)
--- PASS: TestAccAWSElasticacheReplicationGroup_updateNodeSize (3047.94s)
--- PASS: TestAccAWSElasticacheReplicationGroup_tags (918.84s)
--- PASS: TestAccAWSElasticacheCluster_NumCacheNodes_Decrease (1120.19s)
--- PASS: TestAccAWSElasticacheReplicationGroup_ClusterMode_NumNodeGroups (4849.48s)
```

### CODE REVIEW

File: aws/resource_aws_elasticache_cluster.go [hunk 1]
```diff
@@ -425,85 +427,68 @@ func resourceAwsElasticacheClusterRead(d *schema.ResourceData, meta interface{})
 	conn := meta.(*AWSClient).elasticacheconn
 	ignoreTagsConfig := meta.(*AWSClient).IgnoreTagsConfig
 
-	req := &elasticache.DescribeCacheClustersInput{
-		CacheClusterId:    aws.String(d.Id()),
-		ShowCacheNodeInfo: aws.Bool(true),
+	c, err := finder.CacheClusterWithNodeInfoByID(conn, d.Id())
+	if tfresource.NotFound(err) {
```
Comment (bflad) [inline aws/resource_aws_elasticache_cluster.go:431]: Reference: https://github.com/hashicorp/terraform-provider-aws/issues/16796

```suggestion
	if !d.IsNewResource() && tfresource.NotFound(err) {
```

File: aws/resource_aws_elasticache_cluster.go [hunk 2]
```diff
@@ -425,85 +427,68 @@ func resourceAwsElasticacheClusterRead(d *schema.ResourceData, meta interface{})
...
+	}
+
+	d.Set("cluster_id", c.CacheClusterId)
+	d.Set("node_type", c.CacheNodeType)
+	d.Set("num_cache_nodes", c.NumCacheNodes)
+	d.Set("engine", c.Engine)
+	d.Set("engine_version", c.EngineVersion)
+	if c.ConfigurationEndpoint != nil {
+		d.Set("port", c.ConfigurationEndpoint.Port)
+		d.Set("configuration_endpoint", aws.String(fmt.Sprintf("%s:%d", *c.ConfigurationEndpoint.Address, *c.ConfigurationEndpoint.Port)))
```
Comment (bflad) [inline aws/resource_aws_elasticache_cluster.go:447]: Dereference safety:

```suggestion
		d.Set("configuration_endpoint", aws.String(fmt.Sprintf("%s:%d", aws.StringValue(c.ConfigurationEndpoint.Address), aws.Int64Value(c.ConfigurationEndpoint.Port))))
```

File: aws/resource_aws_elasticache_cluster.go [hunk 3]
```diff
@@ -425,85 +427,68 @@ func resourceAwsElasticacheClusterRead(d *schema.ResourceData, meta interface{})
...
+		d.Set("cluster_address", aws.String((*c.ConfigurationEndpoint.Address)))
```
Comment (bflad) [inline aws/resource_aws_elasticache_cluster.go:448]: Dereference safety and extraneous wrapping:

```suggestion
		d.Set("cluster_address", c.ConfigurationEndpoint.Address)
```

File: aws/resource_aws_elasticache_cluster.go [hunk 4]
```diff
@@ -425,85 +427,68 @@ func resourceAwsElasticacheClusterRead(d *schema.ResourceData, meta interface{})
...
+	d.Set("maintenance_window", c.PreferredMaintenanceWindow)
+	d.Set("snapshot_window", c.SnapshotWindow)
+	d.Set("snapshot_retention_limit", c.SnapshotRetentionLimit)
+	if c.NotificationConfiguration != nil {
+		if *c.NotificationConfiguration.TopicStatus == "active" {
+			d.Set("notification_topic_arn", c.NotificationConfiguration.TopicArn)
 		}
+	}
+	d.Set("availability_zone", c.PreferredAvailabilityZone)
+	if *c.PreferredAvailabilityZone == "Multiple" {
```
Comment (bflad) [inline aws/resource_aws_elasticache_cluster.go:472]: Deference safety:

```suggestion
	if aws.StringValue(c.PreferredAvailabilityZone) == "Multiple" {
```

File: aws/resource_aws_elasticache_replication_group.go [hunk 5]
```diff
@@ -941,20 +883,10 @@ func deleteElasticacheReplicationGroup(replicationGroupID string, conn *elastica
...
-		Target:     []string{},
-		Refresh:    cacheReplicationGroupStateRefreshFunc(conn, replicationGroupID, []string{}),
-		Timeout:    40 * time.Minute,
-		MinTimeout: 10 * time.Second,
-		Delay:      30 * time.Second,
-	}
-
-	_, err = stateConf.WaitForState()
+	_, err = waiter.ReplicationGroupDeleted(conn, replicationGroupID, timeout)
 	return err
```
Comment (bflad) [inline aws/resource_aws_elasticache_replication_group.go:890]: Recommend wrapping error:

```suggestion
	if err != nil {
		return fmt.Errorf("error waiting for ElastiCache Replication Group (%s) to delete: %w", err)
	}

	return nil
```

**Review by bflad — APPROVED** [review]
Some minor items, otherwise looks good to me. 🚀

File: aws/resource_aws_elasticache_cluster.go (hunk 3 above)
Comment (gdavison) [reply aws/resource_aws_elasticache_cluster.go:448]: Yikes, that was a copy-paste I didn't check

**Pull request merged**